File: apps/desktop/src/lib/accessorySpecLookup.ts
import type { AccessoryKind, AccessoryIdentifierType } from './productCategory';
import type { ProductModel } from '../services/api/ProductModelService';
import { ProductSpecService } from '../services/api/ProductSpecService';

export interface FetchedAccessorySpec {
  model_name: string;
  model_number: string | null;
  part_number: string | null;
  accessory_kind: AccessoryKind | null;
  color_options: string | null;
  product_image_url: string | null;
  description: string | null;
  notes: string | null;
  source: 'database' | 'gemini';
}

const inflightAccessoryLookups = new Map<string, Promise<FetchedAccessorySpec | null>>();

export function findAccessoryByIdentifier(
  models: ProductModel[],
  identifier: string,
  brandId?: number | null,
): ProductModel | null {
  const normalized = identifier.trim().toLowerCase();
  if (!normalized) return null;

  const pool = brandId
    ? models.filter((model) => model.brand_id === brandId && model.category === 'accessory')
    : models.filter((model) => model.category === 'accessory');

  return (
    pool.find((model) => {
      const modelNumber = model.model_number.trim().toLowerCase();
      const partNumber = model.part_number?.trim().toLowerCase();
      return modelNumber === normalized || partNumber === normalized;
    }) ?? null
  );
}

export async function fetchAccessorySpecFromInternet(
  identifier: string,
  options?: {
    identifierType?: AccessoryIdentifierType;
    brandName?: string;
    modelName?: string;
    forceRefresh?: boolean;
  },
): Promise<FetchedAccessorySpec | null> {
  const normalized = identifier.trim();
  const identifierType = options?.identifierType ?? 'model_number';
  const normalizedBrand = options?.brandName?.trim().toLowerCase() ?? '';
  const inflightKey = `${identifierType}:${normalized.toLowerCase()}|${normalizedBrand}${options?.forceRefresh ? '|refresh' : ''}`;
  const existing = inflightAccessoryLookups.get(inflightKey);
  if (existing) {
    return existing;
  }

  const lookupPromise = (async (): Promise<FetchedAccessorySpec | null> => {
    const result = await ProductSpecService.lookupAccessorySpec({
      identifier: normalized,
      identifier_type: identifierType,
      brand_name: options?.brandName?.trim() || undefined,
      model_name: options?.modelName?.trim() || undefined,
      force_refresh: options?.forceRefresh ?? false,
    });

    if (!result?.model_name) return null;

    return {
      model_name: result.model_name,
      model_number: result.model_number,
      part_number: result.part_number,
      accessory_kind: result.accessory_kind,
      color_options: result.color_options,
      product_image_url: result.product_image_url,
      description: result.description,
      notes: result.notes,
      source: 'gemini',
    };
  })();

  inflightAccessoryLookups.set(inflightKey, lookupPromise);
  try {
    return await lookupPromise;
  } finally {
    if (inflightAccessoryLookups.get(inflightKey) === lookupPromise) {
      inflightAccessoryLookups.delete(inflightKey);
    }
  }
}

export function accessoryToDisplaySpec(model: ProductModel): FetchedAccessorySpec {
  return {
    model_name: model.model_name,
    model_number: model.model_number,
    part_number: model.part_number,
    accessory_kind: model.accessory_kind,
    color_options: model.color_options,
    product_image_url: model.product_image_url,
    description: null,
    notes: model.notes,
    source: 'database',
  };
}

File: apps/mobile_flutter/lib/features/inventory/domain/accessory_spec_lookup.dart
import 'inventory_models.dart';
import 'product_category.dart';

class FetchedAccessorySpec {
  const FetchedAccessorySpec({
    required this.modelName,
    this.modelNumber,
    this.partNumber,
    this.accessoryKind,
    this.colorOptions,
    this.productImageUrl,
    this.description,
    this.notes,
    this.source = AccessorySpecSource.gemini,
  });

  final String modelName;
  final String? modelNumber;
  final String? partNumber;
  final AccessoryKind? accessoryKind;
  final String? colorOptions;
  final String? productImageUrl;
  final String? description;
  final String? notes;
  final AccessorySpecSource source;
}

enum AccessorySpecSource { database, gemini }

ProductModel? findAccessoryByIdentifier(
  List<ProductModel> models,
  String identifier, {
  int? brandId,
}) {
  final normalized = identifier.trim().toLowerCase();
  if (normalized.isEmpty) return null;

  final pool = brandId == null
      ? models.where((model) => model.isAccessory)
      : models.where((model) => model.brandId == brandId && model.isAccessory);

  for (final model in pool) {
    final modelNumber = model.modelNumber.trim().toLowerCase();
    final partNumber = model.partNumber?.trim().toLowerCase();
    if (modelNumber == normalized || partNumber == normalized) {
      return model;
    }
  }
  return null;
}

FetchedAccessorySpec accessoryToDisplaySpec(ProductModel model) {
  return FetchedAccessorySpec(
    modelName: model.modelName,
    modelNumber: model.modelNumber,
    partNumber: model.partNumber,
    accessoryKind: model.accessoryKind,
    colorOptions: model.colorOptions,
    productImageUrl: model.productImageUrl,
    description: null,
    notes: model.notes,
    source: AccessorySpecSource.database,
  );
}

FetchedAccessorySpec fetchedAccessorySpecFromApi(Map<String, dynamic> json) {
  return FetchedAccessorySpec(
    modelName: json['model_name'] as String? ?? '',
    modelNumber: json['model_number'] as String?,
    partNumber: json['part_number'] as String?,
    accessoryKind: accessoryKindFromString(json['accessory_kind'] as String?),
    colorOptions: json['color_options'] as String?,
    productImageUrl: json['product_image_url'] as String?,
    description: json['description'] as String?,
    notes: json['notes'] as String?,
    source: AccessorySpecSource.gemini,
  );
}

final _inflightAccessoryLookups = <String, Future<FetchedAccessorySpec?>>{};

Future<FetchedAccessorySpec?> fetchAccessorySpecFromInternet(
  Future<FetchedAccessorySpec?> Function() lookup, {
  required String identifier,
  required AccessoryIdentifierType identifierType,
  String? brandName,
  bool forceRefresh = false,
}) {
  final normalized = identifier.trim();
  final normalizedBrand = brandName?.trim().toLowerCase() ?? '';
  final inflightKey =
      '${identifierTypeToApi(identifierType)}:${normalized.toLowerCase()}|$normalizedBrand${forceRefresh ? '|refresh' : ''}';
  final existing = _inflightAccessoryLookups[inflightKey];
  if (existing != null) return existing;

  late final Future<FetchedAccessorySpec?> lookupPromise;
  lookupPromise = lookup().whenComplete(() {
    if (_inflightAccessoryLookups[inflightKey] == lookupPromise) {
      _inflightAccessoryLookups.remove(inflightKey);
    }
  });
  _inflightAccessoryLookups[inflightKey] = lookupPromise;
  return lookupPromise;
}
